### PR TITLE
Give S3 bucket owner full control over the published maven artifacts

### LINKED ISSUE #5233: Publishing to S3 Maven: Give bucket owner full control over published files
### Expected Behavior
When publishing artifacts to S3-backed Maven repository, I expect bucket owner (account that owns the bucket) to have full control over the published files, so that the bucket owner can grant read access to the files.

### Current Behavior
The behavior is as expected only if the IAM user/role used to publish artifact belongs to the same account, that owns the bucket. However, if the publishing IAM user/role belongs to another account, the bucket owner by default will not have permissions to read the files. In order for bucket owner to grant read access to the files, the publishing IAM user/role has to grant bucket owner permissions to the published files. For more details, see [Bucket Owner Granting Its Users Permissions to Objects It Does Not Own](https://docs.aws.amazon.com/AmazonS3/latest/dev/example-walkthroughs-managing-access-example3.html).

### Proposed change
When uploading artifacts To S3-backed Maven repository, owner of the bucket should always be granted full access. Since the [MavenArtifactRepository](https://docs.gradle.org/current/dsl/org.gradle.api.artifacts.repositories.MavenArtifactRepository.html) does not seem to provide any way to configure the behavior or does not provide any point, where the configuration could be easily introduced, I propose to change the behavior without option for configuration.

In terms of functionality this should not be a breaking change, as with that modification more principals (bucket owner and object owner) have access to the objects. In terms of security this could be a breaking change, but I have hard time to imagine a use case where the S3 Maven bucket owner should have less permissions then the object owner.

The change can be implemented by adding:

    putObjectRequest.setCannedAcl(CannedAccessControlList.BucketOwnerFullControl)

as line 140 to [S3Client](https://github.com/gradle/gradle/blob/7283def6e4c9773ed92340e57ae5c2033b783c05/subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3Client.java#L142).

### Context
To separate our environments we are using multiple AWS accounts. We also have a root account, where IAM users for employees are defined, so that we can manage users in a single place. The users can access other accounts by assuming roles. The S3 bucket is provisioned on other account than IAM users.

in one account (and can assume roles to control other accounts), while S3 bucket hosting Maven repository is owned by another account. The only other alternative solution I can think of is to migrate the S3 bucket to be owned by the same account where the IAM users are defined.

### PR DESCRIPTION
**Pull request by salvianreynaldi**
Give S3 bucket owner full control over the published maven artifacts,
useful in organizations with multiple AWS accounts, where the S3 bucket might be managed by a team in its own account
fixes #5233

### CODE REVIEW

**Comment by salvianreynaldi** [comment]
@eriwen it's my first time creating a gradle pull request, is this enough?

**Comment by eriwen** [comment]
@salvianreynaldi this is a great start. I appreciate that you've added a test. An integration test would be even better. Is this possible? 

Finally, is there any documentation we can point users to here? Perhaps adding a link in the Gradle docs would be appropriate for users.

**Comment by salvianreynaldi** [comment]
I've try adding some notes to the `Maven Publishing` doc. For the integration test I'm afraid I'm not sure where to start, though. Can somebody help?

**Comment by klesniewski** [comment]
How would you envision the integration test @eriwen? Ideally we would test it with real AWS S3, but do you have an AWS account, against which you run tests in CI? Without it, it is not really possible. We could try to mock AWS S3 (there are mock services available), but:

1. Since it is mock, there is no guarantee that code passing tests will work with real AWS. The mocks are not maintained by AWS.
2. We would essentially test whether AWS S3 SDK library is implemented correctly. I think though it is fair assumption, that the AWS provided library is tested by AWS and will work correctly with S3.

The test @salvianreynaldi provided seems sufficient to me. He verifies that S3 library is called with the required ACL explicitly set. That should prevent others from accidentally breaking this behavior in the future. Perhaps it would be beneficial to extract it as a separate test or to add a line of  explanation (comment pointing to issue). Not sure what the policy is.

What do you think?

@salvianreynaldi, could you sing-off you commits as explained [here](https://github.com/probot/dco#how-it-works)? It is a requirement. Otherwise your changes can't be merged.

**Comment by klesniewski** [comment]
@eriwen ?

I have noticed that CI Status was previously fine, but is now failing. Checking the details though requires logging in to TeamCity. Does it require any special permissions, or can any TeamCity user access it to find out what's the problem?

**Comment by thc202** [comment]
@klesniewski you can choose "Log in as guest".

**Comment by marcphilipp** [comment]
> I have noticed that CI Status was previously fine, but is now failing. Checking the details though requires logging in to TeamCity. Does it require any special permissions, or can any TeamCity user access it to find out what's the problem?

@klesniewski Don't worry about it. It's just there so the Gradle team does not merge a PR while CI is already failing.

@blindpirate This is the second time I'm aware of a contributor was confused by the CI Status check. Do you have any ideas how we can make it more clear to contributors what it means?

**Comment by blindpirate** [comment]
@klesniewski @marcphilipp Thanks for bringing it up.

Firstly, I think we shouldn't put the status here which don't have any TC status yet - If there's no TC checks at all, it seems unnecessary to know if TC is healthy.

Secondly, I think "CI Health"  might be better to describe what's this check is.

WDYT

**Comment by klesniewski** [comment]
Thanks for explanation. I think that putting what was wrote here into Details of "CI Status" check (as proposed by @blindpirate) will be very helpful and won't take much time. There still may be some confusion when it is failing, but at least contributors can find out for themselves that they shouldn't really care and just have to wait.

Regarding the pull request though, can we get it merged?

**Comment by klesniewski** [comment]
It seems that now everything is in place, but CI failed the build. It however seems, that the failure is not due to changes. How can we progress from here?

**Comment by marcphilipp** [comment]
Looks like the build was canceled for some reason... I restarted it:
https://builds.gradle.org/viewQueued.html?itemId=13858239

**Comment by marcphilipp** [comment]
@salvianreynaldi Could you please rebase onto `master` so we can get a green CI build? There have been some changes around Java 7 and TLS v1.2 recently that we need.

**Comment by salvianreynaldi** [comment]
okay @marcphilipp, thanks

**Comment by klesniewski** [comment]
Thanks @salvianreynaldi. Now all checks are green, are we ready to merge @marcphilipp?

**Comment by klesniewski** [comment]
Same outcome. Should it be started again?

**Comment by marcphilipp** [comment]
@klesniewski There are failing tests in the `resourcesS3` subproject (see https://builds.gradle.org/viewLog.html?buildId=13878457).

@salvianreynaldi Are these tests green locally for you?

**Comment by marcphilipp** [comment]
@salvianreynaldi What happened? If you're lacking time to work on this PR, maybe someone else can take over?

**Comment by salvianreynaldi** [comment]
I wrongly force-pushed the branch @marcphilipp 
sorry I hadn't got the time to work on this, but I've tried to fix the integration test, hopefully it does the trick

**Comment by salvianreynaldi** [comment]
btw how should I run test locally? i need to disable `-Werror`, cd-ing to subprojects/resourcs-s3, and run `../../gradlew check` to do it. calling `./gradlew resourcesS3: check` from the root project directory will produce errors not related to the s3 project

**Comment by marcphilipp** [comment]
`./gradlew :resourcesS3:check` should work. Which version of Java are you using? What's the error you get?

**Comment by salvianreynaldi** [comment]
I use OpenJDK 10, and the error is
```
Task :platformPlay:compileJava FAILED
/home/salvian/gradle/subprojects/platform-play/src/main/java/org/gradle/scala/internal/reflect/ScalaListBuffer.java:26: warning: [deprecation] newInstance() in Class has been deprecated
            Object buffer = bufferClass.newInstance();
                                       ^
where T is a type-variable:
    T extends Object declared in class Class
error: warnings found and -Werror specified
1 error
```
that's why I have to disable `-Werror` in `strict-compile.gradle.kts`

I forget what was the error when executing `./gradlew resourcesS3: check`, but I run
```
rm <rootProjectDir>/.gradle -r
./gradlew clean
```
and now `./gradlew resourcesS3: check` runs successfully. I guess it's good to restart the build now

**Comment by marcphilipp** [comment]
Build started: https://builds.gradle.org/viewQueued.html?itemId=14419927&tab=queuedBuildOverviewTab

**Comment by klesniewski** [comment]
Thanks for your effort @salvianreynaldi! Looks like we are finally getting there!

**Comment by salvianreynaldi** [comment]
I've rebased my changes onto latest successful-checks master, https://github.com/gradle/gradle/commit/a5581d16225c14f85c735a4673a28f110e60cb77

**Comment by marcphilipp** [comment]
The build looks good now! https://builds.gradle.org/viewLog.html?buildId=14449853&buildTypeId=Gradle_Check_Stage_BranchBuildAccept_Trigger

File: subprojects/docs/src/docs/userguide/publishing_maven.adoc [hunk 1]
```diff
@@ -123,11 +123,13 @@ The two main things you will want to configure are the repository's:
 
  * URL (required)
  * Name (optional)
- 
+
 You can define multiple repositories as long as they have unique names within the build script. You may also declare one (and only one) repository without a name. That repository will take on an implicit name of "Maven".
 
 You can also configure any authentication details that are required to connect to the repository. See api:org.gradle.api.artifacts.repositories.MavenArtifactRepository[] for more details.
 
+Some organizations may have multiple AWS accounts, e.g. one for each team. The repository bucket's account is often different from the artifact publisher's and consumers'. The bucket owner needs to be able to grant consumer accounts access to the uploaded objects. This is done by adding the `bucket-owner-full-control` link:https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html#canned-acl[Canned ACL] to the uploaded objects. Gradle will do this in every upload. Make sure the publisher has the required IAM permissions (`PutObject` and `PutObjectAcl`). You can read more at link:https://docs.aws.amazon.com/AmazonS3/latest/dev/s3-access-control.html[AWS S3 Access Permissions].
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/publishing_maven.adoc]: This belongs in repository_types.adoc

**Comment by salvianreynaldi** [comment]
right @oehme , as this also apply to Ivy, not just Maven. Is this sufficient?

**Comment by marcphilipp** [comment]
@salvianreynaldi Could you take care of that?

**Comment by salvianreynaldi** [comment]
Do I put it correctly? (in notes.md)

**Comment by klesniewski** [comment]
All green, amazing! I can see it doesn't get automatically merged, so what are the next steps?

**Comment by marcphilipp** [comment]
We're almost there. I started another build because of the rebase: https://builds.gradle.org/viewQueued.html?itemId=14608588&tab=queuedBuildOverviewTab

**Comment by marcphilipp** [comment]
Green build: https://builds.gradle.org/viewLog.html?buildId=14628112&buildTypeId=Gradle_Check_Stage_BranchBuildAccept_Trigger 🎉

**Comment by marcphilipp** [comment]
Thanks, @salvianreynaldi and @klesniewski! 👍

**Comment by klesniewski** [comment]
It was a bit of back and forth, but nevertheless it is now on master! Thank you guys for your effort!

**Pull request merged**